---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noAwaitInLoop
description: Learn more about noAwaitInLoop
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="JavaScript (and super languages)" icon="seti:javascript">
## Summary
:::danger
This rule hasn't been released yet. If you want to try it, you can request and try a preview release of Biome
:::
- Diagnostic Category: [`lint/nursery/noAwaitInLoop`](/reference/diagnostics#diagnostic-category)
- This rule doesn't have a fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- Sources: 
  - Same as [`no-await-in-loop`](https://eslint.org/docs/latest/rules/no-await-in-loop)

:::caution
This rule is part of the [nursery](/linter/rules/#nursery) group.
:::
## Description
Disallow `await` inside loops.

Using `await` in a loop makes your asynchronous operations run one after another instead of all at once. This can slow things down and might cause unhandled errors. Instead, create all the promises together and then wait for them simultaneously using methods like `Promise.all()`.

## Examples

### Invalid

```js
async function invalid() {
    for (const thing of things) {
        const result = await asyncWork();
    }
}
```

<pre class="language-text"><code class="language-text">code-block.js:3:24 <a href="https://biomejs.dev/linter/rules/no-await-in-loop">lint/nursery/noAwaitInLoop</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Avoid using </span><span style="color: lightgreen;"><strong>await</strong></span><span style="color: lightgreen;"> inside loops.</span><br />  <br />    <strong>1 │ </strong>async function invalid() &#123;<br />    <strong>2 │ </strong>    for (const thing of things) &#123;<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>3 │ </strong>        const result = await asyncWork();<br />   <strong>   │ </strong>                       <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>4 │ </strong>    &#125;<br />    <strong>5 │ </strong>&#125;<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Using </span><span style="color: lightgreen;"><strong>await</strong></span><span style="color: lightgreen;"> inside loops might cause performance issues or unintended sequential execution, consider use </span><span style="color: lightgreen;"><strong>Promise.all()</strong></span><span style="color: lightgreen;"> instead.</span><br />  <br /></code></pre>

### Valid

```js
async function valid() {
    await Promise.all(things.map((thing) => asyncWork(thing)))
}
```

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"noAwaitInLoop": "error"
			}
		}
	}
}

```
## Related links

- [Disable a rule](/linter/#disable-a-lint-rule)
- [Configure the rule fix](/linter#configure-the-rule-fix)
- [Rule options](/linter/#rule-options)

</TabItem>
</Tabs>

