---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noDivRegex
description: Learn more about noDivRegex
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="JavaScript (and super languages)" icon="seti:javascript">
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
## Summary
- Rule available since: `v2.3.12`
- Diagnostic Category: [`lint/nursery/noDivRegex`](/reference/diagnostics#diagnostic-category)
- This rule has a [**safe**](/linter/#safe-fixes) fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- Sources: 
  - Same as [`no-div-regex`](https://eslint.org/docs/latest/rules/no-div-regex)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"noDivRegex": "error"
			}
		}
	}
}

```
## Description
Disallow equal signs explicitly at the beginning of regular expressions.

This rule forbids equal signs (`=`) after the slash (`/`) at the beginning of a regular expression literal,
because the characters `/=` can be confused with a division assignment operator.

## Examples

### Invalid

```js
function bar() {
  return /=foo/;
}
```

<pre class="language-text"><code class="language-text">code-block.js:2:10 <a href="https://biomejs.dev/linter/rules/no-div-regex">lint/nursery/noDivRegex</a> <span style="color: #000; background-color: #ddd;"> FIXABLE </span> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Avoid using an equal sign directly after the slash at the beginning of a regular expression literal.</span><br />  <br />    <strong>1 │ </strong>function bar() &#123;<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>2 │ </strong>  return /=foo/;<br />   <strong>   │ </strong>         <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>3 │ </strong>&#125;<br />    <strong>4 │ </strong><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">The characters </span><span style="color: lightgreen;"><strong>/=</strong></span><span style="color: lightgreen;"> can be confused with a division assignment operator. Replace the equal sign (</span><span style="color: lightgreen;"><strong>=</strong></span><span style="color: lightgreen;">) with </span><span style="color: lightgreen;"><strong>[=]</strong></span><span style="color: lightgreen;"> to prevent confusion.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">This rule belongs to the nursery group, which means it is not yet stable and may change in the future. Visit </span><span style="color: lightgreen;"><a href="https://biomejs.dev/linter/#nursery">https://biomejs.dev/linter/#nursery</a></span><span style="color: lightgreen;"> for more information.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Safe fix</span><span style="color: lightgreen;">: </span><span style="color: lightgreen;">Replace with </span><span style="color: lightgreen;"><strong>[=]</strong></span><span style="color: lightgreen;">.</span><br />  <br />  <strong>  2 │ </strong><span style="opacity: 0.8;">·</span><span style="opacity: 0.8;">·</span>return<span style="opacity: 0.8;">·</span>/<span style="color: MediumSeaGreen;">[</span>=<span style="color: MediumSeaGreen;">]</span>foo/;<br />  <strong>    │ </strong>          <span style="color: MediumSeaGreen;">+</span> <span style="color: MediumSeaGreen;">+</span>     <br /></code></pre>

### Valid

```js
function bar() {
  return /[=]foo/;
}
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/src/lint/nursery/no_div_regex.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/tests/specs/nursery/noDivRegex)

</TabItem>
</Tabs>

