---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noDuplicateFields
description: Learn more about noDuplicateFields
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="GraphQL" icon="seti:graphql">
## Summary
- Rule available since: `v1.9.0`
- Diagnostic Category: [`lint/suspicious/noDuplicateFields`](/reference/diagnostics#diagnostic-category)
- This rule is **recommended**, meaning it is enabled by default.
- This rule doesn't have a fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- Sources: 
  - Same as [`@graphql-eslint/no-duplicate-fields`](https://the-guild.dev/graphql/eslint/rules/no-duplicate-fields)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"suspicious": {
				"noDuplicateFields": "error"
			}
		}
	}
}

```
## Description
No duplicated fields in GraphQL operations.

Checks for duplicate fields in selection set, variables in operation definition, or in arguments set of a field.

## Examples

### Invalid

```graphql
query {
  users {
    id
    name
    email
    name
  }
}
```

<pre class="language-text"><code class="language-text">code-block.graphql:6:5 <a href="https://biomejs.dev/linter/rules/no-duplicate-fields">lint/suspicious/noDuplicateFields</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Field &#96;name&#96; defined multiple times.</span><br />  <br />    <strong>4 │ </strong>    name<br />    <strong>5 │ </strong>    email<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>6 │ </strong>    name<br />   <strong>   │ </strong>    <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>7 │ </strong>  &#125;<br />    <strong>8 │ </strong>&#125;<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Remove the duplicated field.</span><br />  <br /></code></pre>

```graphql
query {
  users(
    first: 100,
    after: 10,
    filter: "test",
    first: 50
  ) {
    id
  }
}
```

<pre class="language-text"><code class="language-text">code-block.graphql:6:5 <a href="https://biomejs.dev/linter/rules/no-duplicate-fields">lint/suspicious/noDuplicateFields</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Argument &#96;first&#96; defined multiple times.</span><br />  <br />    <strong>4 │ </strong>    after: 10,<br />    <strong>5 │ </strong>    filter: &quot;test&quot;,<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>6 │ </strong>    first: 50<br />   <strong>   │ </strong>    <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>7 │ </strong>  ) &#123;<br />    <strong>8 │ </strong>    id<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Remove the duplicated argument.</span><br />  <br /></code></pre>

```graphql
query ($v: String, $t: String, $v: String) {
  id
}
```

<pre class="language-text"><code class="language-text">code-block.graphql:1:32 <a href="https://biomejs.dev/linter/rules/no-duplicate-fields">lint/suspicious/noDuplicateFields</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Variable &#96;v&#96; defined multiple times.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 │ </strong>query ($v: String, $t: String, $v: String) &#123;<br />   <strong>   │ </strong>                               <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 │ </strong>  id<br />    <strong>3 │ </strong>&#125;<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Remove the duplicated variable.</span><br />  <br /></code></pre>

### Valid

```graphql
query {
  users {
    id
    name
    email
  }
}
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_graphql_analyze/src/lint/suspicious/no_duplicate_fields.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_graphql_analyze/tests/specs/suspicious/noDuplicateFields)

</TabItem>
</Tabs>

