---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noDuplicateEnumValueNames
description: Learn more about noDuplicateEnumValueNames
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="GraphQL" icon="seti:graphql">
:::note
This rule has been implemented but not released yet. It will be available in the next release.
:::
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
## Summary
- Diagnostic Category: [`lint/nursery/noDuplicateEnumValueNames`](/reference/diagnostics#diagnostic-category)
- This rule doesn't have a fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- Sources: 
  - Same as [`@graphql-eslint/unique-enum-value-names`](https://the-guild.dev/graphql/eslint/rules/unique-enum-value-names)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"noDuplicateEnumValueNames": "error"
			}
		}
	}
}

```
## Description
Require all enum value names to be unique.

A GraphQL enum type is only valid if all its values are uniquely named.
The enum value names are case insensitive, meaning `TEST` & `Test` are seen as the same enum value name.

## Examples

### Invalid

```graphql
enum A {
  TEST
  OTHER
  TEST
}
```

<pre class="language-text"><code class="language-text">code-block.graphql:3:8 <a href="https://biomejs.dev/linter/rules/no-duplicate-enum-value-names">lint/nursery/noDuplicateEnumValueNames</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Duplicate enum value name.</span><br />  <br />    <strong>1 │ </strong>enum A &#123;<br />    <strong>2 │ </strong>  TEST<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>3 │ </strong>  OTHER<br />   <strong>   │ </strong>       <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>4 │ </strong>  TEST<br />   <strong>   │ </strong>  <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>5 │ </strong>&#125;<br />    <strong>6 │ </strong><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">A GraphQL enum type is only valid if all its values are uniquely named. Make sure to name every enum value differently.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">This rule belongs to the nursery group, which means it is not yet stable and may change in the future. Visit </span><span style="color: lightgreen;"><a href="https://biomejs.dev/linter/#nursery">https://biomejs.dev/linter/#nursery</a></span><span style="color: lightgreen;"> for more information.</span><br />  <br /></code></pre>

```graphql
enum B {
  TEST
  TesT
}
```

<pre class="language-text"><code class="language-text">code-block.graphql:2:7 <a href="https://biomejs.dev/linter/rules/no-duplicate-enum-value-names">lint/nursery/noDuplicateEnumValueNames</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Duplicate enum value name.</span><br />  <br />    <strong>1 │ </strong>enum B &#123;<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>2 │ </strong>  TEST<br />   <strong>   │ </strong>      <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>3 │ </strong>  TesT<br />   <strong>   │ </strong>  <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>4 │ </strong>&#125;<br />    <strong>5 │ </strong><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">A GraphQL enum type is only valid if all its values are uniquely named. Make sure to name every enum value differently.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">This rule belongs to the nursery group, which means it is not yet stable and may change in the future. Visit </span><span style="color: lightgreen;"><a href="https://biomejs.dev/linter/#nursery">https://biomejs.dev/linter/#nursery</a></span><span style="color: lightgreen;"> for more information.</span><br />  <br /></code></pre>

### Valid

```graphql
enum A {
  TEST
  OTHER
}
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_graphql_analyze/src/lint/nursery/no_duplicate_enum_value_names.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_graphql_analyze/tests/specs/nursery/noDuplicateEnumValueNames)

</TabItem>
</Tabs>

