---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noFloatingPromises
description: Learn more about noFloatingPromises
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="TypeScript and TSX" icon="seti:typescript">
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
:::note
This rule belongs to the project domain. This means that its activation will activate the Biome Scanner, which might affect the performance. Read more about it in the [documentation page](/linter/domains#project)
:::
## Summary
- Rule available since: `v2.0.0`
- Diagnostic Category: [`lint/nursery/noFloatingPromises`](/reference/diagnostics#diagnostic-category)
- This rule has an [**unsafe**](/linter/#unsafe-fixes) fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- This rule belongs to the following domains:
  - [`project`](/linter/domains#project)
- Sources: 
  - Same as [`@typescript-eslint/no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"noFloatingPromises": "error"
			}
		}
	}
}

```
## Description
Require Promise-like statements to be handled appropriately.

A "floating" `Promise` is one that is created without any code set up to
handle any errors it might throw. Floating Promises can lead to several
issues, including improperly sequenced operations, unhandled Promise
rejections, and other unintended consequences.

This rule will report Promise-valued statements that are not treated in
one of the following ways:

- Calling its `.then()` method with two arguments
- Calling its `.catch()` method with one argument
- `await`-ing it
- `return`-ing it
- `void`-ing it

## Examples

### Invalid

```ts
async function returnsPromise(): Promise<string> {
  return 'value';
}
returnsPromise().then(() => {});
```

```ts
const returnsPromise = async (): Promise<string> => {
  return 'value';
}
async function returnsPromiseInAsyncFunction() {
  returnsPromise().then(() => {});
}
```

```ts
const promise = new Promise((resolve) => resolve('value'));
promise.then(() => { }).finally(() => { });
```

```ts
Promise.all([p1, p2, p3])
```

```ts
class Api {
  async returnsPromise(): Promise<string> {
    return 'value';
  }
  async someMethod() {
    this.returnsPromise();
  }
}
```

```ts
class Parent {
  async returnsPromise(): Promise<string> {
    return 'value';
  }
}

class Child extends Parent {
  async someMethod() {
    this.returnsPromise();
  }
}
```

```ts
class Api {
  async returnsPromise(): Promise<string> {
    return 'value';
  }
}
const api = new Api();
api.returnsPromise().then(() => {}).finally(() => {});
```

```ts
const obj = {
  async returnsPromise(): Promise<string> {
    return 'value';
  },
};

obj.returnsPromise();
```

```ts
type Props = {
  returnsPromise: () => Promise<void>;
};

async function testCallingReturnsPromise(props: Props) {
  props.returnsPromise();
}
```

### Valid

```ts
async function returnsPromise(): Promise<string> {
  return 'value';
}

await returnsPromise();

void returnsPromise();

// Calling .then() with two arguments
returnsPromise().then(
  () => {},
  () => {},
);

// Calling .catch() with one argument
returnsPromise().catch(() => {});

await Promise.all([p1, p2, p3])

class Api {
  async returnsPromise(): Promise<string> {
    return 'value';
  }
  async someMethod() {
    await this.returnsPromise();
  }
}
```

```ts
type Props = {
  returnsPromise: () => Promise<void>;
};

async function testCallingReturnsPromise(props: Props) {
  return props.returnsPromise();
}
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/src/lint/nursery/no_floating_promises.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/tests/specs/nursery/noFloatingPromises)

</TabItem>
</Tabs>

