---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noFloatingPromises
description: Learn more about noFloatingPromises
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="TypeScript and TSX" icon="seti:typescript">
## Summary
:::note
This rule is currently only available in the [2.0 beta release](/blog/biome-v2-0-beta).
:::
- Diagnostic Category: [`lint/nursery/noFloatingPromises`](/reference/diagnostics#diagnostic-category)
- This rule has an [**unsafe**](/linter/#unsafe-fixes) fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- Sources: 
  - Same as [`@typescript-eslint/no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises)

:::caution
This rule is part of the [nursery](/linter/rules/#nursery) group.
:::
## Description
Require Promise-like statements to be handled appropriately.

A "floating" `Promise` is one that is created without any code set up to handle any errors it might throw.
Floating Promises can lead to several issues, including improperly sequenced operations, unhandled Promise rejections, and other unintended consequences.

This rule will report Promise-valued statements that are not treated in one of the following ways:

- Calling its `.then()` method with two arguments
- Calling its `.catch()` method with one argument
- `await`ing it
- `return`ing it
- `void`ing it

:::caution

## Important notes

This rule is a work in progress, and is only partially implemented.
Progress is being tracked in the following GitHub issue: https://github.com/biomejs/biome/issues/3187
:::

## Examples

### Invalid

```ts
async function returnsPromise(): Promise<string> {
  return 'value';
}
returnsPromise().then(() => {});
```

```ts
const returnsPromise = async (): Promise<string> => {
  return 'value';
}
async function returnsPromiseInAsyncFunction() {
  returnsPromise().then(() => {});
}
```

```ts
const promise = new Promise((resolve) => resolve('value'));
promise.then(() => { }).finally(() => { });
```

```ts
Promise.all([p1, p2, p3])
```

```ts
class Api {
  async returnsPromise(): Promise<string> {
    return 'value';
  }
  async someMethod() {
    this.returnsPromise();
  }
}
```

```ts
class Parent {
  async returnsPromise(): Promise<string> {
    return 'value';
  }
}

class Child extends Parent {
  async someMethod() {
    this.returnsPromise();
  }
}
```

```ts
class Api {
  async returnsPromise(): Promise<string> {
    return 'value';
  }
}
const api = new Api();
api.returnsPromise().then(() => {}).finally(() => {});
```

```ts
const obj = {
  async returnsPromise(): Promise<string> {
    return 'value';
  },
};

obj.returnsPromise();
```

```ts
type Props = {
  returnsPromise: () => Promise<void>;
};

async function testCallingReturnsPromise(props: Props) {
  props.returnsPromise();
}
```

### Valid

```ts
async function returnsPromise(): Promise<string> {
  return 'value';
}

await returnsPromise();

void returnsPromise();

// Calling .then() with two arguments
returnsPromise().then(
  () => {},
  () => {},
);

// Calling .catch() with one argument
returnsPromise().catch(() => {});

await Promise.all([p1, p2, p3])

class Api {
  async returnsPromise(): Promise<string> {
    return 'value';
  }
  async someMethod() {
    await this.returnsPromise();
  }
}
```

```ts
type Props = {
  returnsPromise: () => Promise<void>;
};

async function testCallingReturnsPromise(props: Props) {
  return props.returnsPromise();
}
```

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"noFloatingPromises": "error"
			}
		}
	}
}

```
## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the rule fix](/linter#configure-the-rule-fix)
- [Rule options](/linter/#rule-options)

</TabItem>
</Tabs>

