---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noIncrementDecrement
description: Learn more about noIncrementDecrement
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="JavaScript (and super languages)" icon="seti:javascript">
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
## Summary
- Rule available since: `v2.3.2`
- Diagnostic Category: [`lint/nursery/noIncrementDecrement`](/reference/diagnostics#diagnostic-category)
- This rule doesn't have a fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- Sources: 
  - Same as [`no-plusplus`](https://eslint.org/docs/latest/rules/no-plusplus)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"noIncrementDecrement": "error"
			}
		}
	}
}

```
## Description
Disallows the usage of the unary operators ++ and --.

Because the unary ++ and -- operators are subject to automatic semicolon insertion, differences in whitespace can change semantics of source code.

```js
let i = 10;
let j = 20;

i ++
j
// i = 11, j = 20
```

<pre class="language-text"><code class="language-text">code-block.js:4:1 <a href="https://biomejs.dev/linter/rules/no-increment-decrement">lint/nursery/noIncrementDecrement</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Unexpected use of increment/decrement unary operator.</span><br />  <br />    <strong>2 │ </strong>let j = 20;<br />    <strong>3 │ </strong><br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>4 │ </strong>i ++<br />   <strong>   │ </strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>5 │ </strong>j<br />    <strong>6 │ </strong>// i = 11, j = 20<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">The unary ++ and -- operators are subject to automatic semicolon insertion, differences in whitespace can change semantics of source code. Instead use += or -=.</span><br />  <br /></code></pre>

```js
let i = 10;
let j = 20;

i
++
j
// i = 10, j = 21
```

<pre class="language-text"><code class="language-text">code-block.js:5:1 <a href="https://biomejs.dev/linter/rules/no-increment-decrement">lint/nursery/noIncrementDecrement</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Unexpected use of increment/decrement unary operator.</span><br />  <br />    <strong>4 │ </strong>i<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>5 │ </strong>++<br />   <strong>   │ </strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>6 │ </strong>j<br />   <strong>   │ </strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>7 │ </strong>// i = 10, j = 21<br />    <strong>8 │ </strong><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">The unary ++ and -- operators are subject to automatic semicolon insertion, differences in whitespace can change semantics of source code. Instead use += or -=.</span><br />  <br /></code></pre>

## Examples

### Invalid

```js
let foo = 0;
foo++;
```

<pre class="language-text"><code class="language-text">code-block.js:2:1 <a href="https://biomejs.dev/linter/rules/no-increment-decrement">lint/nursery/noIncrementDecrement</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Unexpected use of increment/decrement unary operator.</span><br />  <br />    <strong>1 │ </strong>let foo = 0;<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>2 │ </strong>foo++;<br />   <strong>   │ </strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>3 │ </strong><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">The unary ++ and -- operators are subject to automatic semicolon insertion, differences in whitespace can change semantics of source code. Instead use += or -=.</span><br />  <br /></code></pre>

```js
let bar = 42;
bar--;
```

<pre class="language-text"><code class="language-text">code-block.js:2:1 <a href="https://biomejs.dev/linter/rules/no-increment-decrement">lint/nursery/noIncrementDecrement</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Unexpected use of increment/decrement unary operator.</span><br />  <br />    <strong>1 │ </strong>let bar = 42;<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>2 │ </strong>bar--;<br />   <strong>   │ </strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>3 │ </strong><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">The unary ++ and -- operators are subject to automatic semicolon insertion, differences in whitespace can change semantics of source code. Instead use += or -=.</span><br />  <br /></code></pre>

```js
for (let i = 0; i < 10; i++) {
    doSomething(i);
}
```

<pre class="language-text"><code class="language-text">code-block.js:1:25 <a href="https://biomejs.dev/linter/rules/no-increment-decrement">lint/nursery/noIncrementDecrement</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Unexpected use of increment/decrement unary operator.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 │ </strong>for (let i = 0; i &lt; 10; i++) &#123;<br />   <strong>   │ </strong>                        <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 │ </strong>    doSomething(i);<br />    <strong>3 │ </strong>&#125;<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">The unary ++ and -- operators are subject to automatic semicolon insertion, differences in whitespace can change semantics of source code. Instead use += or -=.</span><br />  <br /></code></pre>

```js
for (let i = 0; i < 10;) {
    doSomething(i);
    i++;
}
```

<pre class="language-text"><code class="language-text">code-block.js:3:5 <a href="https://biomejs.dev/linter/rules/no-increment-decrement">lint/nursery/noIncrementDecrement</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Unexpected use of increment/decrement unary operator.</span><br />  <br />    <strong>1 │ </strong>for (let i = 0; i &lt; 10;) &#123;<br />    <strong>2 │ </strong>    doSomething(i);<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>3 │ </strong>    i++;<br />   <strong>   │ </strong>    <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>4 │ </strong>&#125;<br />    <strong>5 │ </strong><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">The unary ++ and -- operators are subject to automatic semicolon insertion, differences in whitespace can change semantics of source code. Instead use += or -=.</span><br />  <br /></code></pre>

### Valid

```js
let foo = 0;
foo += 1;
```

```js
let bar = 42;
bar -= 1;
```

```js
for (let i = 0; i < 10; i += 1) {
    doSomething(i);
}
```

```js
for (let i = 0; i < 10;) {
    doSomething(i);
    i += 1;
}
```

## Options

### `allowForLoopAfterthoughts`

Allows unary operators ++ and -- in the afterthought (final expression) of a for loop.

Default `false`

```json title='biome.json'
{
	"linter": {
		"rules": {
			"nursery": {
				"noIncrementDecrement": {
					"options": {
						"allowForLoopAfterthoughts": true
					}
				}
			}
		}
	}
}

```

#### Invalid

```js
for (let i = 0; i < j; j = i++) {
    doSomething(i, j);
}
```

<pre class="language-text"><code class="language-text">code-block.js:1:28 <a href="https://biomejs.dev/linter/rules/no-increment-decrement">lint/nursery/noIncrementDecrement</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Unexpected use of increment/decrement unary operator.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 │ </strong>for (let i = 0; i &lt; j; j = i++) &#123;<br />   <strong>   │ </strong>                           <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 │ </strong>    doSomething(i, j);<br />    <strong>3 │ </strong>&#125;<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">The unary ++ and -- operators are subject to automatic semicolon insertion, differences in whitespace can change semantics of source code. Instead use += or -=.</span><br />  <br /></code></pre>

```js
for (let i = 10; i--;) {
    doSomething(i);
}
```

<pre class="language-text"><code class="language-text">code-block.js:1:18 <a href="https://biomejs.dev/linter/rules/no-increment-decrement">lint/nursery/noIncrementDecrement</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Unexpected use of increment/decrement unary operator.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 │ </strong>for (let i = 10; i--;) &#123;<br />   <strong>   │ </strong>                 <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 │ </strong>    doSomething(i);<br />    <strong>3 │ </strong>&#125;<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">The unary ++ and -- operators are subject to automatic semicolon insertion, differences in whitespace can change semantics of source code. Instead use += or -=.</span><br />  <br /></code></pre>

```js
for (let i = 0; i < 10;) i++;
```

<pre class="language-text"><code class="language-text">code-block.js:1:26 <a href="https://biomejs.dev/linter/rules/no-increment-decrement">lint/nursery/noIncrementDecrement</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Unexpected use of increment/decrement unary operator.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 │ </strong>for (let i = 0; i &lt; 10;) i++;<br />   <strong>   │ </strong>                         <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 │ </strong><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">The unary ++ and -- operators are subject to automatic semicolon insertion, differences in whitespace can change semantics of source code. Instead use += or -=.</span><br />  <br /></code></pre>

#### Valid

```js
for (let i = 0; i < 10; i++) {
    doSomething(i);
}
```

```js
for (let i = 0, j = l; i < l; i++, j--) {
    doSomething(i, j);
}
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/src/lint/nursery/no_increment_decrement.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/tests/specs/nursery/noIncrementDecrement)

</TabItem>
</Tabs>

