---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noMisleadingCharacterClass
description: |
  Disallow characters made with multiple code points in character class syntax.
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

**Diagnostic Category: `lint/suspicious/noMisleadingCharacterClass`**
<Tabs>
<TabItem label="JavaScript (and super languages)" icon="seti:javascript">
**Since**: `v1.5.0`
:::note
- This rule is **recommended**. A [diagnostic error](/reference/diagnostics#error) will appear when linting your code.
- This rule has a **safe** fix.
:::

Sources: 
- Same as: <a href="https://eslint.org/docs/latest/rules/no-misleading-character-class" target="_blank"><code>no-misleading-character-class</code></a>

Disallow characters made with multiple code points in character class syntax.

Unicode includes the characters which are made with multiple code points. e.g. AÌ, ğŸ‡¯ğŸ‡µ, ğŸ‘¨â€ğŸ‘©â€ğŸ‘¦.
A RegExp character class `/[abc]/` cannot handle characters with multiple code points.
For example, the character `â‡ï¸` consists of two code points: `â‡` (U+2747) and `VARIATION SELECTOR-16` (U+FE0F).
If this character is in a RegExp character class, it will match to either `â‡` or `VARIATION SELECTOR-16` rather than `â‡ï¸`.
This rule reports the regular expressions which include multiple code point characters in character class syntax.

## Examples

### Invalid

```js
/^[AÌ]$/u;
```

<pre class="language-text"><code class="language-text">code-block.js:1:4 <a href="https://biomejs.dev/linter/rules/no-misleading-character-class">lint/suspicious/noMisleadingCharacterClass</a> â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”<br /><br />  <strong><span style="color: Tomato;">âœ–</span></strong> <span style="color: Tomato;">A character class cannot match a character and a combining character.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 â”‚ </strong>/^[AÌ]$/u;<br />   <strong>   â”‚ </strong>   <strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 â”‚ </strong><br />  <br />  <strong><span style="color: lightgreen;">â„¹</span></strong> <span style="color: lightgreen;">A character and a combining character forms a new character. Replace the character class with an alternation.</span><br />  <br /></code></pre>

```js
/^[â‡ï¸]$/u;
```

<pre class="language-text"><code class="language-text">code-block.js:1:4 <a href="https://biomejs.dev/linter/rules/no-misleading-character-class">lint/suspicious/noMisleadingCharacterClass</a> â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”<br /><br />  <strong><span style="color: Tomato;">âœ–</span></strong> <span style="color: Tomato;">A character class cannot match a character and a combining character.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 â”‚ </strong>/^[â‡ï¸]$/u;<br />   <strong>   â”‚ </strong>   <strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 â”‚ </strong><br />  <br />  <strong><span style="color: lightgreen;">â„¹</span></strong> <span style="color: lightgreen;">A character and a combining character forms a new character. Replace the character class with an alternation.</span><br />  <br /></code></pre>

```js
/^[ğŸ‘¶ğŸ»]$/u;
```

<pre class="language-text"><code class="language-text">code-block.js:1:4 <a href="https://biomejs.dev/linter/rules/no-misleading-character-class">lint/suspicious/noMisleadingCharacterClass</a> â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”<br /><br />  <strong><span style="color: Tomato;">âœ–</span></strong> <span style="color: Tomato;">A character class cannot match an emoji with a skin tone modifier.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 â”‚ </strong>/^[ğŸ‘¶ğŸ»]$/u;<br />   <strong>   â”‚ </strong>   <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 â”‚ </strong><br />  <br />  <strong><span style="color: lightgreen;">â„¹</span></strong> <span style="color: lightgreen;">Replace the character class with an alternation.</span><br />  <br /></code></pre>

```js
/^[ğŸ‡¯ğŸ‡µ]$/u;
```

<pre class="language-text"><code class="language-text">code-block.js:1:4 <a href="https://biomejs.dev/linter/rules/no-misleading-character-class">lint/suspicious/noMisleadingCharacterClass</a> â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”<br /><br />  <strong><span style="color: Tomato;">âœ–</span></strong> <span style="color: Tomato;">A character class cannot match a pair of regional indicator symbols.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 â”‚ </strong>/^[ğŸ‡¯ğŸ‡µ]$/u;<br />   <strong>   â”‚ </strong>   <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 â”‚ </strong><br />  <br />  <strong><span style="color: lightgreen;">â„¹</span></strong> <span style="color: lightgreen;">A pair of regional indicator symbols encodes a country code. Replace the character class with an alternation.</span><br />  <br /></code></pre>

```js
/^[ğŸ‘¨â€ğŸ‘©â€ğŸ‘¦]$/u;
```

<pre class="language-text"><code class="language-text">code-block.js:1:4 <a href="https://biomejs.dev/linter/rules/no-misleading-character-class">lint/suspicious/noMisleadingCharacterClass</a> â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”<br /><br />  <strong><span style="color: Tomato;">âœ–</span></strong> <span style="color: Tomato;">A character class cannot match a joined character sequence.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 â”‚ </strong>/^[ğŸ‘¨â€ğŸ‘©â€ğŸ‘¦]$/u;<br />   <strong>   â”‚ </strong>   <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 â”‚ </strong><br />  <br />  <strong><span style="color: lightgreen;">â„¹</span></strong> <span style="color: lightgreen;">A zero width joiner composes several emojis into a new one. Replace the character class with an alternation.</span><br />  <br /></code></pre>

```js
/^[ğŸ‘]$/; // surrogate pair without u flag
```

<pre class="language-text"><code class="language-text">code-block.js:1:4 <a href="https://biomejs.dev/linter/rules/no-misleading-character-class">lint/suspicious/noMisleadingCharacterClass</a> <span style="color: #000; background-color: #ddd;"> FIXABLE </span> â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”<br /><br />  <strong><span style="color: Tomato;">âœ–</span></strong> <span style="color: Tomato;">A character class cannot match a surrogate pair. Add the 'u' unicode flag to match against them.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 â”‚ </strong>/^[ğŸ‘]$/; // surrogate pair without u flag<br />   <strong>   â”‚ </strong>   <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 â”‚ </strong><br />  <br />  <strong><span style="color: lightgreen;">â„¹</span></strong> <span style="color: lightgreen;">A surrogate pair forms a single codepoint, but is encoded as a pair of two characters. Without the unicode flag, the regex matches a single surrogate character.</span><br />  <br />  <strong><span style="color: lightgreen;">â„¹</span></strong> <span style="color: lightgreen;">Safe fix</span><span style="color: lightgreen;">: </span><span style="color: lightgreen;">Add unicode </span><span style="color: lightgreen;"><strong>u</strong></span><span style="color: lightgreen;"> flag to regex</span><br />  <br />  <strong>  1 â”‚ </strong>/^[ğŸ‘]$/<span style="color: MediumSeaGreen;">u</span>;<span style="opacity: 0.8;">Â·</span>//<span style="opacity: 0.8;">Â·</span>surrogate<span style="opacity: 0.8;">Â·</span>pair<span style="opacity: 0.8;">Â·</span>without<span style="opacity: 0.8;">Â·</span>u<span style="opacity: 0.8;">Â·</span>flag<br />  <strong>    â”‚ </strong>        <span style="color: MediumSeaGreen;">+</span>                                  <br /></code></pre>

### Valid

```js
/^[abc]$/;
/^[ğŸ‘]$/u;
/^[\q{ğŸ‘¶ğŸ»}]$/v;
```

## Related links

- [Disable a rule](/linter/#disable-a-lint-rule)
- [Configure the rule fix](/linter#configure-the-rule-fix)
- [Rule options](/linter/#rule-options)

</TabItem>
</Tabs>

