---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noQwikUseVisibleTask
description: Learn more about noQwikUseVisibleTask
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="JSX and TSX" icon="seti:javascript">
## Summary
:::note
This rule is currently only available in the [2.0 beta release](/blog/biome-v2-0-beta).
:::
- Diagnostic Category: [`lint/nursery/noQwikUseVisibleTask`](/reference/diagnostics#diagnostic-category)
- This rule doesn't have a fix.
- The default severity of this rule is [**warning**](/reference/diagnostics#warning).
- This rule belongs to the following domains:
  - [`qwik`](/linter/domains#qwik)
- Sources: 
  - Same as [`qwik/no-use-visible-task`](https://github.com/BuilderIO/eslint-plugin-qwik/blob/main/docs/rules/no-use-visible-task.md)

:::caution
This rule is part of the [nursery](/linter/rules/#nursery) group.
:::
## Description
Disallow `useVisibleTask$()` functions in Qwik components.

Prevents hydration-blocking operations that hurt Qwik's resumability.
See [Qwik Tasks Documentation](https://qwik.dev/docs/components/tasks/) for proper alternatives.

## Examples

### Invalid

```js
useVisibleTask$(() => {
  console.log('Component is visible');
});
```

<pre class="language-text"><code class="language-text">code-block.js:1:1 <a href="https://biomejs.dev/linter/rules/no-qwik-use-visible-task">lint/nursery/noQwikUseVisibleTask</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: Orange;">⚠</span></strong> <span style="color: Orange;">Avoid </span><span style="color: Orange;"><strong>useVisibleTask$</strong></span><span style="color: Orange;"> for non-interactive initialization</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 │ </strong>useVisibleTask$(() =&gt; &#123;<br />   <strong>   │ </strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 │ </strong>  console.log('Component is visible');<br />    <strong>3 │ </strong>&#125;);<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">This hook executes immediately on component mount without user interaction, potentially:<br /></span>    <span style="color: lightgreen;">- Hurting performance (blocking hydration)<br /></span>    <span style="color: lightgreen;">- Causing layout shifts (CLS)<br /></span>    <span style="color: lightgreen;">- Breaking SSR compatibility</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Check the </span><span style="color: lightgreen;"><a href="https://qwik.dev/docs/components/tasks/">Qwik documentation</a></span><span style="color: lightgreen;"> for suitable alternatives.</span><br />  <br /></code></pre>

### Valid

```js
useTask$(() => {
  console.log('Task executed');
});
```

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"noQwikUseVisibleTask": "error"
			}
		}
	}
}

```
## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)

</TabItem>
</Tabs>

