---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noQuickfixBiome
description: Learn more about noQuickfixBiome
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="JSON (and super languages)" icon="seti:json">
## Summary
- Rule available since: `v2.1.3`
- Diagnostic Category: [`lint/suspicious/noQuickfixBiome`](/reference/diagnostics#diagnostic-category)
- This rule is **recommended**, meaning it is enabled by default.
- This rule has a [**safe**](/linter/#safe-fixes) fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"suspicious": {
				"noQuickfixBiome": "error"
			}
		}
	}
}

```
## Description
Disallow the use if `quickfix.biome` inside editor settings file.

The code action `quickfix.biome` can be harmful because it instructs the editors
to apply the code fix of lint rules and code actions atomically. If multiple rules or
actions apply a code fix to the same code span, the editor will emit invalid code.

The rule targets specifically VSCode settings and Zed settings. Specifically, paths that end with:

- `.vscode/settings.json`
- `Code/User/settings.json`
- `.zed/settings.json`
- `zed/settings.json`

## Examples

### Invalid

```json
{
    "quickfix.biome": "explicit"
}
```

### Valid

```json
{
    "source.fixAll.biome": "explicit"
}
```

## Options

The following options are available

### `additionalPaths`

It's possible to specify a list of JSON paths, if your editor uses a JSON file setting that isn't supported natively by the rule.

If your editor uses, for example, a file called `.myEditor/file.json`, you can add `".myEditor/file.json"` to the list.
**The rule checks if the file ends with the given paths**.

```json title='biome.json'
{
	"linter": {
		"rules": {
			"suspicious": {
				"noQuickfixBiome": {
					"options": {
						"additionalPaths": [
							".myEditor/file.json"
						]
					}
				}
			}
		}
	}
}

```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_json_analyze/src/lint/suspicious/no_quickfix_biome.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_json_analyze/tests/specs/suspicious/noQuickfixBiome)

</TabItem>
</Tabs>

