---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noRestrictedImports
description: |
  Disallow specified modules when loaded by import or require.
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

**Diagnostic Category: `lint/nursery/noRestrictedImports`**
<Tabs>
<TabItem label="JavaScript (and super languages)" icon="seti:javascript">
**Since**: `v1.6.0`
:::caution
This rule is part of the [nursery](/linter/rules/#nursery) group.
:::

Sources: 
- Same as: <a href="https://eslint.org/docs/latest/rules/no-restricted-imports" target="_blank"><code>no-restricted-imports</code></a>
- Same as: <a href="https://typescript-eslint.io/rules/no-restricted-imports" target="_blank"><code>@typescript-eslint/no-restricted-imports</code></a>

Disallow specified modules when loaded by import or require.

## Options

```options
{
    "options": {
        "paths": {
            "lodash": "Using lodash is not encouraged",
            "underscore": "Using underscore is not encouraged"
        }
    }
}
```

## Examples

### Invalid

```use_options
import "lodash";
import "allowed-import";
```

<pre class="language-text"><code class="language-text"></code></pre>

```use_options
const underscore = await import("underscore");
```

<pre class="language-text"><code class="language-text"></code></pre>

```use_options
const lodash = require("lodash");
```

<pre class="language-text"><code class="language-text"></code></pre>

### Valid

```use_options
import "allowed-import";
const myImport = await import("allowed-import");
const myImport = require("allowed-import");
```

## Related links

- [Disable a rule](/linter/#disable-a-lint-rule)
- [Configure the rule fix](/linter#configure-the-rule-fix)
- [Rule options](/linter/#rule-options)

</TabItem>
</Tabs>

