---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noUndeclaredDependencies
description: |
  Disallow the use of dependencies that aren\'t specified in the `package.json`.
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

**Diagnostic Category: `lint/correctness/noUndeclaredDependencies`**
<Tabs>
<TabItem label="JavaScript (and super languages)" icon="seti:javascript">
**Since**: `v1.6.0`
## Description
Disallow the use of dependencies that aren't specified in the `package.json`.

Indirect dependencies will trigger the rule because they aren't declared in the `package.json`.
This means that if the package `@org/foo` has a dependency on `lodash`, and then you use
`import "lodash"` somewhere in your project, the rule will trigger a diagnostic for this import.

The rule ignores imports that are not valid package names.
This includes internal imports that start with `#` and `@/` and imports with a protocol such as `node:`, `bun:`, `jsr:`, `https:`.

To ensure that Visual Studio Code uses relative imports when it automatically imports a variable,
you may set [`javascript.preferences.importModuleSpecifier` and `typescript.preferences.importModuleSpecifier`](https://code.visualstudio.com/docs/getstarted/settings) to `relative`.

## Examples

### Invalid

```js
import "vite";
```

### Valid

```js
import { A } from "./local.js";
```

```js
import assert from "node:assert";
```

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"correctness": {
				"noUndeclaredDependencies": "error"
			}
		}
	}
}

```
## Related links

- [Disable a rule](/linter/#disable-a-lint-rule)
- [Configure the rule fix](/linter#configure-the-rule-fix)
- [Rule options](/linter/#rule-options)

</TabItem>
</Tabs>

