---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noUnknownAtRule
description: |
  Disallow unknown at-rules.
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

**Diagnostic Category: `lint/nursery/noUnknownAtRule`**
<Tabs>
<TabItem label="CSS" icon="seti:css">
**Since**: `vnext`
:::caution
This rule is part of the [nursery](/linter/rules/#nursery) group.
:::

Sources: 
- Same as: <a href="https://github.com/stylelint/stylelint/blob/main/lib/rules/at-rule-no-unknown/README.md" target="_blank"><code>stylelint/at-rule-no-unknown</code></a>

Disallow unknown at-rules.

For details on known at-rules, see the [MDN web docs](https://developer.mozilla.org/en-US/docs/Web/CSS/At-rule).

## Examples

### Invalid

```css
@uNkNoWn {}
```

<pre class="language-text"><code class="language-text">code-block.css:1:2 <a href="https://biomejs.dev/linter/rules/no-unknown-at-rule">lint/nursery/noUnknownAtRule</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: Tomato;">✖</span></strong> <span style="color: Tomato;">Unexpected unknown at-rule: </span><span style="color: Tomato;"><strong>uNkNoWn</strong></span><span style="color: Tomato;"> </span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 │ </strong>@uNkNoWn &#123;&#125;<br />   <strong>   │ </strong> <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 │ </strong><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;"><strong>uNkNoWn</strong></span><span style="color: lightgreen;"> is not a standard CSS at-rule, which may lead to unexpected styling results or failure to interpret the styles as intended.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">See </span><span style="color: lightgreen;"><a href="https://developer.mozilla.org/en-US/docs/Web/CSS/At-rule">MDN web docs</a></span><span style="color: lightgreen;"> for a known list of at-rules.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">To fix this issue, consider removing the unknown at-rule.</span><br />  <br /></code></pre>

```css
@unknown-at-rule {
  font-size: 14px;
}
```

<pre class="language-text"><code class="language-text">code-block.css:1:2 <a href="https://biomejs.dev/linter/rules/no-unknown-at-rule">lint/nursery/noUnknownAtRule</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: Tomato;">✖</span></strong> <span style="color: Tomato;">Unexpected unknown at-rule: </span><span style="color: Tomato;"><strong>unknown-at-rule</strong></span><span style="color: Tomato;"> </span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 │ </strong>@unknown-at-rule &#123;<br />   <strong>   │ </strong> <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 │ </strong>  font-size: 14px;<br />    <strong>3 │ </strong>&#125;<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;"><strong>unknown-at-rule</strong></span><span style="color: lightgreen;"> is not a standard CSS at-rule, which may lead to unexpected styling results or failure to interpret the styles as intended.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">See </span><span style="color: lightgreen;"><a href="https://developer.mozilla.org/en-US/docs/Web/CSS/At-rule">MDN web docs</a></span><span style="color: lightgreen;"> for a known list of at-rules.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">To fix this issue, consider removing the unknown at-rule.</span><br />  <br /></code></pre>

### Valid

```css
@charset 'UTF-8';
```

```css
@media (max-width: 960px) {
  body {
    font-size: 13px;
  }
}
```

## Related links

- [Disable a rule](/linter/#disable-a-lint-rule)
- [Configure the rule fix](/linter#configure-the-rule-fix)
- [Rule options](/linter/#rule-options)

</TabItem>
</Tabs>

