---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noUnresolvedImports
description: Learn more about noUnresolvedImports
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="JavaScript (and super languages)" icon="seti:javascript">
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
:::note
This rule belongs to the project domain. This means that its activation will activate the Biome Scanner, which might affect the performance. Read more about it in the [documentation page](/linter/domains#project)
:::
## Summary
- Rule available since: `v2.0.0`
- Diagnostic Category: [`lint/nursery/noUnresolvedImports`](/reference/diagnostics#diagnostic-category)
- This rule doesn't have a fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- This rule belongs to the following domains:
  - [`project`](/linter/domains#project)
- Sources: 
  - Inspired from [`import/named`](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/named.md)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"noUnresolvedImports": "error"
			}
		}
	}
}

```
## Description
Warn when importing non-existing exports.

Importing a non-existing export is an error at runtime or build time.
Biome can detect such incorrect imports and report errors for them.

Note that if you use TypeScript, you probably don't want to use this
rule, since TypeScript already performs such checks for you.

## Known Limitations

- This rule does not validate imports through dynamic `import()`
expressions or CommonJS `require()` calls.

## Examples

### Invalid

**`foo.js`**

```js
export function foo() {};
```

**`bar.js`**

```js
// Attempt to import symbol with a typo:
import { fooo } from "./foo.js";
```

### Valid

**`bar.js`**

```js
// Fixed typo:
import { foo } from "./foo.js";
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/src/lint/nursery/no_unresolved_imports.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/tests/specs/nursery/noUnresolvedImports)

</TabItem>
</Tabs>

