---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: noValueAtRule
description: Learn more about noValueAtRule
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="CSS" icon="seti:css">
## Summary
- Rule available since: `v1.8.0`
- Diagnostic Category: [`lint/style/noValueAtRule`](/reference/diagnostics#diagnostic-category)
- This rule isn't recommended, so you need to enable it.
- This rule doesn't have a fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"style": {
				"noValueAtRule": "error"
			}
		}
	}
}

```
## Description
Disallow use of `@value` rule in CSS modules.

Use of CSS variables is recommended instead of `@value` rule.

## Examples

### Invalid

```css title='example.module.css'
@value red: #FF0000;
```

<pre class="language-text"><code class="language-text"></code></pre>

### Valid

```css title='example.module.css'
:root {
  --red: #FF0000
}

p {
  background-color: var(--red);
}
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_css_analyze/src/lint/style/no_value_at_rule.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_css_analyze/tests/specs/style/noValueAtRule)

</TabItem>
</Tabs>

