---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: useVueConsistentVOnStyle
description: Learn more about useVueConsistentVOnStyle
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="HTML" icon="seti:html">
:::note
This rule has been implemented but not released yet. It will be available in the next release.
:::
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
## Summary
- Diagnostic Category: [`lint/nursery/useVueConsistentVOnStyle`](/reference/diagnostics#diagnostic-category)
- This rule has an [**unsafe**](/linter/#unsafe-fixes) fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- This rule belongs to the following domains:
  - [`vue`](/linter/domains#vue)
- Sources: 
  - Same as [`vue/v-on-style`](https://eslint.vuejs.org/rules/v-on-style)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"useVueConsistentVOnStyle": "error"
			}
		}
	}
}

```
## Description
Enforce a consistent style for `v-on` in Vue templates.

## Examples

### Invalid

```vue
<div v-on:click="onClick" />
```

<pre class="language-text"><code class="language-text"></code></pre>

### Valid

```vue
<div @click="onClick" />
```

## Options

### `style`

Configures the preferred directive style. Accepts `"shorthand"` or `"longhand"`. Default: `"shorthand"`.

```json title='biome.json'
{
	"linter": {
		"rules": {
			"nursery": {
				"useVueConsistentVOnStyle": {
					"options": {
						"style": "longhand"
					}
				}
			}
		}
	}
}

```

#### Invalid

```vue
<div @click="onClick" />
```

<pre class="language-text"><code class="language-text"></code></pre>

#### Valid

```vue
<div v-on:click="onClick" />
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/src/lint/nursery/use_vue_consistent_v_on_style.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/tests/specs/nursery/useVueConsistentVOnStyle)

</TabItem>
</Tabs>

