---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: useVueValidVElseIf
description: Learn more about useVueValidVElseIf
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="HTML" icon="seti:html">
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
## Summary
- Rule available since: `v2.3.6`
- Diagnostic Category: [`lint/nursery/useVueValidVElseIf`](/reference/diagnostics#diagnostic-category)
- This rule doesn't have a fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- This rule belongs to the following domains:
  - [`vue`](/linter/domains#vue)
- Sources: 
  - Same as [`vue/valid-v-else-if`](https://eslint.vuejs.org/rules/valid-v-else-if)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"useVueValidVElseIf": "error"
			}
		}
	}
}

```
## Description
Enforce valid `v-else-if` directives.

Biome flags these cases:

- Has an argument: `<div v-else-if:arg="b"></div>`.
- Has modifiers: `<div v-else-if.mod="b"></div>`.
- Missing value: `<div v-else-if></div>`.
- Not preceded by a sibling with `v-if`/`v-else-if`.
- On the same element as `v-if` or `v-else`.

## Examples

### Invalid

```vue
<div v-if="a"></div><div v-else-if:arg="b"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-if="a"></div><div v-else-if.mod="b"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-if="a"></div><div v-else-if></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-else-if="b"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-if="a" v-else-if="b"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

### Valid

```vue
<div v-if="a"></div><div v-else-if="b"></div><div v-else></div>
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/src/lint/nursery/use_vue_valid_v_else_if.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/tests/specs/nursery/useVueValidVElseIf)

</TabItem>
</Tabs>

