---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: useVueValidVIf
description: Learn more about useVueValidVIf
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="HTML" icon="seti:html">
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
## Summary
- Rule available since: `v2.3.6`
- Diagnostic Category: [`lint/nursery/useVueValidVIf`](/reference/diagnostics#diagnostic-category)
- This rule doesn't have a fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- This rule belongs to the following domains:
  - [`vue`](/linter/domains#vue)
- Sources: 
  - Same as [`vue/valid-v-if`](https://eslint.vuejs.org/rules/valid-v-if)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"useVueValidVIf": "error"
			}
		}
	}
}

```
## Description
Enforces valid `v-if` usage for Vue templates.

This rule reports `v-if` directives in following cases:

- The directive has an argument. E.g. `<div v-if:aaa="foo"></div>`
- The directive has a modifier. E.g. `<div v-if.bbb="foo"></div>`
- The directive does not have an attribute value. E.g. `<div v-if></div>`
- The same element also has `v-else` or `v-else-if`. E.g. `<div v-if="foo" v-else></div>`

## Examples

### Invalid

```vue
<div v-if:aaa="foo"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-if.bbb="foo"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-if></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-if="foo" v-else></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-if="foo" v-else-if="bar"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

### Valid

```vue
<div v-if="ok"></div>
```

```vue
<div v-if="a < b"></div>
```

```vue
<div v-if="a"></div>
<div v-else-if="b"></div>
<div v-else></div>
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/src/lint/nursery/use_vue_valid_v_if.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/tests/specs/nursery/useVueValidVIf)

</TabItem>
</Tabs>

