---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: useVueValidVElse
description: Learn more about useVueValidVElse
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="HTML" icon="seti:html">
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
## Summary
- Rule available since: `v2.3.6`
- Diagnostic Category: [`lint/nursery/useVueValidVElse`](/reference/diagnostics#diagnostic-category)
- This rule doesn't have a fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- This rule belongs to the following domains:
  - [`vue`](/linter/domains#vue)
- Sources: 
  - Same as [`vue/valid-v-else`](https://eslint.vuejs.org/rules/valid-v-else)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"useVueValidVElse": "error"
			}
		}
	}
}

```
## Description
Enforce valid usage of v-else.

This rule reports v-else directives in the following cases:

- The directive has an argument. E.g. `<div v-if="foo"></div><div v-else:aaa></div>`
- The directive has a modifier. E.g. `<div v-if="foo"></div><div v-else.bbb></div>`
- The directive has an attribute value. E.g. `<div v-if="foo"></div><div v-else="bar"></div>`
- The directive is on elements where the previous element doesn't have `v-if`/`v-else-if` directives. E.g. `<div v-else></div>`
- The directive is on elements which have `v-if`/`v-else-if` directives. E.g. `<div v-if="foo" v-else></div>`

## Examples

### Invalid

```vue
<div v-else:arg></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-else.mod></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-else="value"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-else></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-if="foo" v-else></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

### Valid

```vue
<div v-if="foo"></div>
<div v-else></div>
```

```vue
<div v-if="foo"></div>
<div v-else-if="bar"></div>
<div v-else></div>
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/src/lint/nursery/use_vue_valid_v_else.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/tests/specs/nursery/useVueValidVElse)

</TabItem>
</Tabs>

