---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: useConsistentMemberAccessibility
description: |
  Require consistent accessibility modifiers on class properties and methods.
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

**Diagnostic Category: `lint/nursery/useConsistentMemberAccessibility`**
<Tabs>
<TabItem label="TypeScript and TSX" icon="seti:typescript">
**Since**: `v1.9.0`
:::caution
This rule is part of the [nursery](/linter/rules/#nursery) group.
:::

Sources: 
- Same as: <a href="https://typescript-eslint.io/rules/explicit-member-accessibility" target="_blank"><code>@typescript-eslint/explicit-member-accessibility</code></a>

Require consistent accessibility modifiers on class properties and methods.

TypeScript allows placing explicit `public`, `protected`, and `private` accessibility modifiers in front of class members.
The modifiers exist solely in the type system and just serve to describe who is allowed to access those members.
Leaving off accessibility modifiers makes for less code to read and write. Members are public by default.

However, adding in consistent accessibility modifiers can be helpful in codebases with many classes for enforcing proper privacy of members.
Some developers also find it preferable for code readability to keep member publicity explicit.

## Examples

### Invalid

#### `"accessibility": "noPublic"` (default value)

Use the following configuration to disallow all explicit `public` modifiers:

```options
{
    "options": {
        "accessibility": "noPublic"
    }
}
```

The following patterns are considered incorrect code with `noPublic`:

```use_options
class Animal {
  public constructor(breed, name) {
    // ...
  }
}
```

<pre class="language-text"><code class="language-text"></code></pre>

```use_options
class Animal {
  constructor(
    public breed,
    name,
  ) {
    // ...
  }
}
```

<pre class="language-text"><code class="language-text"></code></pre>

```use_options
class Animal {
  public animalName: string;
}
```

<pre class="language-text"><code class="language-text"></code></pre>

```use_options
class Pet {
  public get name(): string {
    return this.animalName;
  }
}
```

<pre class="language-text"><code class="language-text"></code></pre>

```use_options
class Pet {
  public set name(value: string) {
    this.animalName = value;
  }
}
```

<pre class="language-text"><code class="language-text"></code></pre>

```use_options
class Dog {
  public walk() {
    // ...
  }
}
```

<pre class="language-text"><code class="language-text"></code></pre>

#### `"accessibility": "explicit"`

Use the following configuration to enforce the presence of explicit modifiers wherever possible:

```options
{
    "options": {
        "accessibility": "explicit"
    }
}
```

The following patterns are considered incorrect code with `accessibility` set to `explicit`:

```use_options
class Animal {
  constructor( // Invalid: Missing accessibility modifier
    public breed,
    name,
  ) {
    this.animalName = name;
  }
  private animalName: string; // OK: Modifier must be present
  public get name(): string { // OK: Modifier must be present
    return this.animalName;
  }
  public set name(value: string) { // OK: Modifier must be present
    this.animalName = value;
  }
  protected walk() { // OK: Modifier must be present
    // ...
  }
}
```

<pre class="language-text"><code class="language-text"></code></pre>

#### `"accessibility": "none"`

Use the following configuration to disallow all explicit visibility modifiers:

```options
{
    "options": {
        "accessibility": "none"
    }
}
```

The following patterns are considered incorrect code with `accessibility` set to `none`:

```use_options
class Animal {
  protected constructor(breed, name) {
    // ...
  }
}
```

<pre class="language-text"><code class="language-text"></code></pre>

```use_options
class Animal {
  constructor(
    protected breed,
    name,
  ) {
    // ...
  }
}
```

<pre class="language-text"><code class="language-text"></code></pre>

```use_options
class Animal {
  private animalName: string;
}
```

<pre class="language-text"><code class="language-text"></code></pre>

```use_options
class Animal {
  protected get name(): string {
    return this.animalName;
  }
}
```

<pre class="language-text"><code class="language-text"></code></pre>

```use_options
class Pet {
  private set name(value: string) {
    this.animalName = value;
  }
}
```

<pre class="language-text"><code class="language-text"></code></pre>

```use_options
class Dog {
  public walk() {
    // ...
  }
}
```

<pre class="language-text"><code class="language-text"></code></pre>

### Valid

The following patterns are considered correct code with the default options `noPublic`:

```options
{
    "options": {
        "accessibility": "noPublic"
    }
}
```

```use_options
class Animal {
  constructor(
    private breed,
    name,
  ) {
    this.animalName = name;
  }
  private animalName: string; // Property
  get name(): string {
    // get accessor
    return this.animalName;
  }
  set name(value: string) {
    // set accessor
    this.animalName = value;
  }
  protected walk() {
    // method
  }
}
```

The following patterns are considered correct code with the accessibility set to `explicit`:

```options
{
    "options": {
        "accessibility": "explicit"
    }
}
```

```use_options
class Animal {
  public constructor(
    public breed,
    name,
  ) {
    // Parameter property and constructor
    this.animalName = name;
  }
  private animalName: string; // Property
  public get name(): string {
    // get accessor
    return this.animalName;
  }
  public set name(value: string) {
    // set accessor
    this.animalName = value;
  }
  protected walk() {
    // method
  }
}
```

The following patterns are considered correct code with the accessibility set to `none`:

```options
{
    "options": {
        "accessibility": "none"
    }
}
```

```use_options
class Animal {
  constructor(
    breed,
    name,
  ) {
    // Parameter property and constructor
    this.name = name;
  }
  animalName: string; // Property
  get name(): string {
    // get accessor
    return this.animalName;
  }
  set name(value: string) {
    // set accessor
    this.animalName = value;
  }
  walk() {
    // method
  }
}
```

## Options

The rule supports the following options:

```options
{
    "options": {
        "accessibility": "explicit"
    }
}
```

### `accessibility`

This option determines the required accessibility modifiers on class properties and methods.
It can be set to one of the following values:

- `noPublic` - forbid the use of public (a safe fix will remove it).
- `explicit` - requires an accessibility modifier for every member that allows that (a safe fix will add public).
- `none` - forbid all accessibility modifiers (public, protected, private).

**Default:** `noPublic`

## Related links

- [Disable a rule](/linter/#disable-a-lint-rule)
- [Configure the rule fix](/linter#configure-the-rule-fix)
- [Rule options](/linter/#rule-options)

</TabItem>
</Tabs>

