---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: useVueValidVText
description: Learn more about useVueValidVText
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="HTML" icon="seti:html">
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
## Summary
- Rule available since: `v2.3.7`
- Diagnostic Category: [`lint/nursery/useVueValidVText`](/reference/diagnostics#diagnostic-category)
- This rule doesn't have a fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- This rule belongs to the following domains:
  - [`vue`](/linter/domains#vue)
- Sources: 
  - Same as [`vue/valid-v-text`](https://eslint.vuejs.org/rules/valid-v-text)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"useVueValidVText": "error"
			}
		}
	}
}

```
## Description
Enforce valid `v-text` Vue directives.

This rule reports `v-text` directives in the following cases:

- The directive has an argument. E.g. `<div v-text:aaa></div>`
- The directive has any modifiers. E.g. `<div v-text.bbb></div>`
- The directive does not have a value. E.g. `<div v-text></div>`

## Examples

### Invalid

```vue
<div v-text />
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-text:aaa="foo"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-text.bbb="foo"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

### Valid

```vue
<div v-text="foo" />
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/src/lint/nursery/use_vue_valid_v_text.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/tests/specs/nursery/useVueValidVText)

</TabItem>
</Tabs>

