---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: useExhaustiveDependencies
description: |
  Enforce all dependencies are correctly specified in a React hook.
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

**Diagnostic Category: `lint/correctness/useExhaustiveDependencies`**
<Tabs>
<TabItem label="JSX and TSX" icon="seti:javascript">
**Since**: `v1.0.0`
:::note
- This rule is recommended by Biome. A diagnostic error will appear when linting your code.
:::

Sources: 
- Same as: <a href="https://github.com/facebook/react/blob/main/packages/eslint-plugin-react-hooks/README.md" target="_blank"><code>react-hooks/exhaustive-deps</code></a>

## Description
Enforce all dependencies are correctly specified in a React hook.

_This rule should be used only in **React** projects._

This rule is a port of the rule [react-hooks/exhaustive-deps](https://legacy.reactjs.org/docs/hooks-rules.html#eslint-plugin), and it's meant to target projects that uses React.

If your project _doesn't_ use React (or Preact), **you shouldn't use this rule**.

The rule will inspect the following **known** hooks:

- `useEffect`
- `useLayoutEffect`
- `useInsertionEffect`
- `useCallback`
- `useMemo`
- `useImperativeHandle`
- `useState`
- `useReducer`
- `useRef`
- `useDebugValue`
- `useDeferredValue`
- `useTransition`

If you want to add more hooks to the rule, check the [options](#options).

## Examples

### Invalid

```js
import { useEffect } from "react";

function component() {
    let a = 1;
    useEffect(() => {
        console.log(a);
    }, []);
}
```

<pre class="language-text"><code class="language-text">code-block.js:5:5 <a href="https://biomejs.dev/linter/rules/use-exhaustive-dependencies">lint/correctness/useExhaustiveDependencies</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: Tomato;">✖</span></strong> <span style="color: Tomato;">This hook does not specify all of its dependencies: a</span><br />  <br />    <strong>3 │ </strong>function component() &#123;<br />    <strong>4 │ </strong>    let a = 1;<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>5 │ </strong>    useEffect(() =&gt; &#123;<br />   <strong>   │ </strong>    <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>6 │ </strong>        console.log(a);<br />    <strong>7 │ </strong>    &#125;, []);<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">This dependency is not specified in the hook dependency list.</span><br />  <br />    <strong>4 │ </strong>    let a = 1;<br />    <strong>5 │ </strong>    useEffect(() =&gt; &#123;<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>6 │ </strong>        console.log(a);<br />   <strong>   │ </strong>                    <strong><span style="color: Tomato;">^</span></strong><br />    <strong>7 │ </strong>    &#125;, []);<br />    <strong>8 │ </strong>&#125;<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Either include it or remove the dependency array</span><br />  <br /></code></pre>

```js
import { useEffect } from "react";

function component() {
    let b = 1;
    useEffect(() => {
    }, [b]);
}
```

<pre class="language-text"><code class="language-text">code-block.js:5:5 <a href="https://biomejs.dev/linter/rules/use-exhaustive-dependencies">lint/correctness/useExhaustiveDependencies</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: Tomato;">✖</span></strong> <span style="color: Tomato;">This hook specifies more dependencies than necessary: b</span><br />  <br />    <strong>3 │ </strong>function component() &#123;<br />    <strong>4 │ </strong>    let b = 1;<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>5 │ </strong>    useEffect(() =&gt; &#123;<br />   <strong>   │ </strong>    <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>6 │ </strong>    &#125;, [b]);<br />    <strong>7 │ </strong>&#125;<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">This dependency can be removed from the list.</span><br />  <br />    <strong>4 │ </strong>    let b = 1;<br />    <strong>5 │ </strong>    useEffect(() =&gt; &#123;<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>6 │ </strong>    &#125;, [b]);<br />   <strong>   │ </strong>        <strong><span style="color: Tomato;">^</span></strong><br />    <strong>7 │ </strong>&#125;<br />    <strong>8 │ </strong><br />  <br /></code></pre>

```js
import { useEffect, useState } from "react";

function component() {
    const [name, setName] = useState();
    useEffect(() => {
        console.log(name);
        setName("");
    }, [name, setName]);
}
```

<pre class="language-text"><code class="language-text">code-block.js:5:5 <a href="https://biomejs.dev/linter/rules/use-exhaustive-dependencies">lint/correctness/useExhaustiveDependencies</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: Tomato;">✖</span></strong> <span style="color: Tomato;">This hook specifies more dependencies than necessary: setName</span><br />  <br />    <strong>3 │ </strong>function component() &#123;<br />    <strong>4 │ </strong>    const [name, setName] = useState();<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>5 │ </strong>    useEffect(() =&gt; &#123;<br />   <strong>   │ </strong>    <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>6 │ </strong>        console.log(name);<br />    <strong>7 │ </strong>        setName(&quot;&quot;);<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">This dependency can be removed from the list.</span><br />  <br />     <strong>6 │ </strong>        console.log(name);<br />     <strong>7 │ </strong>        setName(&quot;&quot;);<br />   <strong><span style="color: Tomato;">&gt;</span></strong> <strong>8 │ </strong>    &#125;, [name, setName]);<br />    <strong>   │ </strong>              <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />     <strong>9 │ </strong>&#125;<br />    <strong>10 │ </strong><br />  <br /></code></pre>

```js
import { useEffect } from "react";

function component() {
    let a = 1;
    const b = a + 1;
    useEffect(() => {
        console.log(b);
    }, []);
}
```

<pre class="language-text"><code class="language-text">code-block.js:6:5 <a href="https://biomejs.dev/linter/rules/use-exhaustive-dependencies">lint/correctness/useExhaustiveDependencies</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: Tomato;">✖</span></strong> <span style="color: Tomato;">This hook does not specify all of its dependencies: b</span><br />  <br />    <strong>4 │ </strong>    let a = 1;<br />    <strong>5 │ </strong>    const b = a + 1;<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>6 │ </strong>    useEffect(() =&gt; &#123;<br />   <strong>   │ </strong>    <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>7 │ </strong>        console.log(b);<br />    <strong>8 │ </strong>    &#125;, []);<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">This dependency is not specified in the hook dependency list.</span><br />  <br />    <strong>5 │ </strong>    const b = a + 1;<br />    <strong>6 │ </strong>    useEffect(() =&gt; &#123;<br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>7 │ </strong>        console.log(b);<br />   <strong>   │ </strong>                    <strong><span style="color: Tomato;">^</span></strong><br />    <strong>8 │ </strong>    &#125;, []);<br />    <strong>9 │ </strong>&#125;<br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Either include it or remove the dependency array</span><br />  <br /></code></pre>

### Valid

```js
import { useEffect } from "react";

function component() {
    let a = 1;
    useEffect(() => {
        console.log(a);
    }, [a]);
}
```

```js
import { useEffect } from "react";

function component() {
    const a = 1;
    useEffect(() => {
        console.log(a);
    });
}
```

```js
import { useEffect, useState } from "react";

function component() {
    const [name, setName] = useState();
    useEffect(() => {
        console.log(name);
        setName("");
    }, [name]);
}
```

```js
import { useEffect } from "react";
let outer = false;
function component() {
    useEffect(() => {
        outer = true;
    }, []);
}
```

## Ignoring a specific dependency

Sometimes you may wish to ignore a diagnostic about a specific
dependency without disabling _all_ linting for that hook. To do so,
you may specify the name of a specific dependency between parentheses,
like this:

```js
import { useEffect } from "react";

function component() {
    let a = 1;
    // biome-ignore lint/correctness/useExhaustiveDependencies(a): <explanation>
    useEffect(() => {
        console.log(a);
    }, []);
}
```

If you wish to ignore multiple dependencies, you can add multiple
comments and add a reason for each.

## Options

Allows to specify custom hooks - from libraries or internal projects -
for which dependencies should be checked and/or which are known to have
stable return values.

### Validating dependencies

For every hook for which you want the dependencies to be validated, you
should specify the index of the closure and the index of the
dependencies array to validate against.

#### Example

```json
{
    "//": "...",
    "options": {
        "hooks": [
            { "name": "useLocation", "closureIndex": 0, "dependenciesIndex": 1},
            { "name": "useQuery", "closureIndex": 1, "dependenciesIndex": 0}
        ]
    }
}
```

Given the previous example, your hooks can be used like this:

```js
function Foo() {
    const location = useLocation(() => {}, []);
    const query = useQuery([], () => {});
}
```

### Stable results

When a hook is known to have a stable return value (its identity doesn't
change across invocations), that value doesn't need to be specified in
dependency arrays. For example, setters returned by React's `useState`
hook always have the same identity and should be omitted as such.

You can configure custom hooks that return stable results in one of
three ways:

- `"stableResult": true` -- marks the return value as stable. An example
of a React hook that would be configured like this is `useRef()`.
- `"stableResult": [1]` -- expects the return value to be an array and
marks the given index or indices to be stable. An example of a React
hook that would be configured like this is `useState()`.
- `"stableResult": 1` -- shorthand for `"stableResult": [1]`.

#### Example

```json
{
    "//": "...",
    "options": {
        "hooks": [
            { "name": "useDispatch", "stableResult": true }
        ]
    }
}
```

With this configuration, the following is valid:

```js
const dispatch = useDispatch();
// No need to list `dispatch` as dependency:
const doAction = useCallback(() => dispatch(someAction()), []);
```

## Preact support

This rule recognizes rules imported from `preact/compat` and
`preact/hooks` and applies the same rules as for React hooks.

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"correctness": {
				"useExhaustiveDependencies": "error"
			}
		}
	}
}

```
## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the rule fix](/linter#configure-the-rule-fix)
- [Rule options](/linter/#rule-options)

</TabItem>
</Tabs>

