---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: useVueValidTemplateRoot
description: Learn more about useVueValidTemplateRoot
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="HTML" icon="seti:html">
:::note
This rule has been implemented but not released yet. It will be available in the next release.
:::
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
## Summary
- Diagnostic Category: [`lint/nursery/useVueValidTemplateRoot`](/reference/diagnostics#diagnostic-category)
- This rule has an [**unsafe**](/linter/#unsafe-fixes) fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- This rule belongs to the following domains:
  - [`vue`](/linter/domains#vue)
- Sources: 
  - Same as [`vue/valid-template-root`](https://eslint.vuejs.org/rules/valid-template-root)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"useVueValidTemplateRoot": "error"
			}
		}
	}
}

```
## Description
Enforce valid Vue `<template>` root usage.

This rule reports only root-level `<template>` elements. If the
`<template>` has a `src` attribute, the element must be empty. Otherwise,
the element must contain content.

## Examples

### Invalid

```vue
<template src="./foo.html">content</template>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<template></template>
```

<pre class="language-text"><code class="language-text"></code></pre>

### Valid

```vue
<template>content</template>
```

```vue
<template src="./foo.html"></template>
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/src/lint/nursery/use_vue_valid_template_root.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/tests/specs/nursery/useVueValidTemplateRoot)

</TabItem>
</Tabs>

