---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: useComponentExportOnlyModules
description: |
  Enforce declaring components only within modules that export React Components exclusively.
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

**Diagnostic Category: `lint/nursery/useComponentExportOnlyModules`**
<Tabs>
<TabItem label="JSX and TSX" icon="seti:javascript">
**Since**: `v1.9.2`
:::caution
This rule is part of the [nursery](/linter/rules/#nursery) group.
:::

Sources: 
- Inspired from: <a href="https://github.com/ArnaudBarre/eslint-plugin-react-refresh" target="_blank"><code>react-refresh/only-export-components</code></a>

Enforce declaring components only within modules that export React Components exclusively.

This is necessary to enable the [`React Fast Refresh`](https://github.com/facebook/react/tree/main/packages/react-refresh) feature, which improves development efficiency.
The determination of whether something is a component depends on naming conventions.
Components should be written in [`PascalCase`](https://en.wikipedia.org/wiki/Camel_case) and regular functions in [`camelCase`](https://en.wikipedia.org/wiki/Camel_case).
If the framework already has established conventions, consider optionally specifying exceptions.

## Examples

### Invalid

```jsx
export const foo = () => {};
export const Bar = () => <></>;
```

<pre class="language-text"><code class="language-text">code-block.jsx:1:14 <a href="https://biomejs.dev/linter/rules/use-components-only-module">lint/nursery/useComponentExportOnlyModules</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: Orange;">⚠</span></strong> <span style="color: Orange;">Exporting a non-component with components is not allowed.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 │ </strong>export const foo = () =&gt; &#123;&#125;;<br />   <strong>   │ </strong>             <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 │ </strong>export const Bar = () =&gt; &lt;&gt;&lt;/&gt;;<br />    <strong>3 │ </strong><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;"><a href="https://github.com/facebook/react/tree/main/packages/react-refresh">Fast Refresh</a></span><span style="color: lightgreen;"> only works when a file only exports components.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Consider separating non-component exports into a new file.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">If it is a component, it may not be following the variable naming conventions.</span><br />  <br /></code></pre>

```jsx
const Tab = () => {};
export const tabs = [<Tab />, <Tab />];
```

<pre class="language-text"><code class="language-text">code-block.jsx:1:7 <a href="https://biomejs.dev/linter/rules/use-components-only-module">lint/nursery/useComponentExportOnlyModules</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: Orange;">⚠</span></strong> <span style="color: Orange;">Components should be exported.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 │ </strong>const Tab = () =&gt; &#123;&#125;;<br />   <strong>   │ </strong>      <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 │ </strong>export const tabs = [&lt;Tab /&gt;, &lt;Tab /&gt;];<br />    <strong>3 │ </strong><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;"><a href="https://github.com/facebook/react/tree/main/packages/react-refresh">Fast Refresh</a></span><span style="color: lightgreen;"> only works when a file only exports components.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Consider separating component exports into a new file.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">If it is not a component, it may not be following the variable naming conventions.</span><br />  <br /></code></pre>

```jsx
const App = () => {}
createRoot(document.getElementById("root")).render(<App />);
```

<pre class="language-text"><code class="language-text">code-block.jsx:1:7 <a href="https://biomejs.dev/linter/rules/use-components-only-module">lint/nursery/useComponentExportOnlyModules</a> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━<br /><br />  <strong><span style="color: Orange;">⚠</span></strong> <span style="color: Orange;">Unexported components are not allowed.</span><br />  <br />  <strong><span style="color: Tomato;">&gt;</span></strong> <strong>1 │ </strong>const App = () =&gt; &#123;&#125;<br />   <strong>   │ </strong>      <strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><strong><span style="color: Tomato;">^</span></strong><br />    <strong>2 │ </strong>createRoot(document.getElementById(&quot;root&quot;)).render(&lt;App /&gt;);<br />    <strong>3 │ </strong><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;"><a href="https://github.com/facebook/react/tree/main/packages/react-refresh">Fast Refresh</a></span><span style="color: lightgreen;"> only works when a file only exports components.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">Consider separating component exports into a new file.</span><br />  <br />  <strong><span style="color: lightgreen;">ℹ</span></strong> <span style="color: lightgreen;">If it is not a component, it may not be following the variable naming conventions.</span><br />  <br /></code></pre>

### Valid

```jsx
export default function Foo() {
    return <></>;
}
```

```jsx
const foo = () => {};
export const Bar = () => <></>;
```

```jsx
import { App } from "./App";
createRoot(document.getElementById("root")).render(<App />);
```

Functions that return standard React components are also permitted.

```jsx
import { memo } from 'react';
const Component = () => <></>
export default memo(Component);
```

## Options

### `allowConstantExport`

Some tools, such as [Vite](https://vitejs.dev/), allow exporting constants along with components. By enabling the following, the rule will support the pattern.

```json
{
    "//": "...",
    "options":{
        "allowConstantExport" : true
    }
}
```

### `allowExportNames`

If you use a framework that handles [Hot Module Replacement(HMR)](https://remix.run/docs/en/main/discussion/hot-module-replacement) of some specific exports, you can use this option to avoid warning for them.

Example for [Remix](https://remix.run/docs/en/main/discussion/hot-module-replacement#supported-exports):

```json
{
    "//": "...",
    "options":{
        "allowExportNames": ["json", "loader", "headers", "meta", "links", "scripts"]
    }
}
```

## Related links

- [Disable a rule](/linter/#disable-a-lint-rule)
- [Configure the rule fix](/linter#configure-the-rule-fix)
- [Rule options](/linter/#rule-options)

</TabItem>
</Tabs>

