---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: useBiomeIgnoreFolder
description: Learn more about useBiomeIgnoreFolder
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="JSON (and super languages)" icon="seti:json">
:::note
This rule has been implemented but not released yet. It will be available in the next release.
:::
## Summary
- Diagnostic Category: [`lint/suspicious/useBiomeIgnoreFolder`](/reference/diagnostics#diagnostic-category)
- This rule is **recommended**, which means is enabled by default.
- This rule has a [**safe**](/linter/#safe-fixes) fix.
- The default severity of this rule is [**warning**](/reference/diagnostics#warning).
## Description
Promotes the correct usage for ignoring folders in the configuration file.

Starting Biome v2.2, ignoring folders doesn't require the use of the trailing `/**`.
When using the pattern `/**`, you tell Biome to ignore **all files** inside a folder, but the folder is still crawled. This pattern
can lead to poor performance, especially if the folder contains many files.

If the intention is to ignore specific files inside a folder, the trailing pattern `/**` shouldn't be used.

## Examples

### Invalid

```json
{
    "files": {
        "includes": ["**", "!dist/**", "!**/dist/**"]
    }
}
```

### Valid

```json
{
    "files": {
        "includes": ["**", "!dist", "!**/dist"]
    }
}
```

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"suspicious": {
				"useBiomeIgnoreFolder": "error"
			}
		}
	}
}

```
## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)

</TabItem>
</Tabs>

