---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: useVueValidVHtml
description: Learn more about useVueValidVHtml
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="HTML" icon="seti:html">
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
## Summary
- Rule available since: `v2.3.6`
- Diagnostic Category: [`lint/nursery/useVueValidVHtml`](/reference/diagnostics#diagnostic-category)
- This rule doesn't have a fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- This rule belongs to the following domains:
  - [`vue`](/linter/domains#vue)
- Sources: 
  - Same as [`vue/valid-v-html`](https://eslint.vuejs.org/rules/valid-v-html)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"useVueValidVHtml": "error"
			}
		}
	}
}

```
## Description
Enforce valid `v-html` directives.

This rule reports v-html directives in the following cases:

- The directive has an argument. E.g. `<div v-html:aaa></div>`
- The directive has a modifier. E.g. `<div v-html.bbb></div>`
- The directive does not have an attribute value. E.g. `<div v-html></div>`

## Examples

### Invalid

```vue
<div v-html:aaa="foo"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-html.bbb="foo"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

```vue
<div v-html></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

### Valid

```vue
<div v-html="htmlContent"></div>
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/src/lint/nursery/use_vue_valid_v_html.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/tests/specs/nursery/useVueValidVHtml)

</TabItem>
</Tabs>

