---
# Don't modify this file manually. This file is auto generated from source, and you will lose your changes next time the website is built.
# Head to the `biomejs/biome` repository, and modify the source code in there.

title: useVueValidVOnce
description: Learn more about useVueValidVOnce
---
import { Tabs, TabItem } from '@astrojs/starlight/components';

<Tabs>
<TabItem label="HTML" icon="seti:html">
:::note
This rule has been implemented but not released yet. It will be available in the next release.
:::
:::caution
This rule is part of the [nursery](/linter/#nursery) group. This means that it is experimental and the behavior can change at any time.
:::
## Summary
- Diagnostic Category: [`lint/nursery/useVueValidVOnce`](/reference/diagnostics#diagnostic-category)
- This rule has an [**unsafe**](/linter/#unsafe-fixes) fix.
- The default severity of this rule is [**information**](/reference/diagnostics#information).
- This rule belongs to the following domains:
  - [`vue`](/linter/domains#vue)
- Sources: 
  - Same as [`vue/valid-v-once`](https://eslint.vuejs.org/rules/valid-v-once)

## How to configure
```json title="biome.json"
{
	"linter": {
		"rules": {
			"nursery": {
				"useVueValidVOnce": "error"
			}
		}
	}
}

```
## Description
Enforce valid `v-once` Vue directives.

## Examples

### Invalid

This rule reports `v-once` directives in the following cases:

The directive has an argument so it is invalid.

```vue
<div v-once:arg></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

The directive has a modifier so it is invalid.

```vue
<div v-once.mod></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

The directive has a value so it is invalid.

```vue
<div v-once="value"></div>
```

<pre class="language-text"><code class="language-text"></code></pre>

### Valid

```vue
<div v-once></div>
```

## Related links

- [Disable a rule](/linter/#disable-a-rule)
- [Configure the code fix](/linter#configure-the-code-fix)
- [Rule options](/linter/#rule-options)
- [Source Code](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/src/lint/nursery/use_vue_valid_v_once.rs)
- [Test Cases](https://github.com/biomejs/biome/blob/main/crates/biome_html_analyze/tests/specs/nursery/useVueValidVOnce)

</TabItem>
</Tabs>

